### PR TITLE
Vary Cache: Validate auth keys and cookie values

### LINKED ISSUE #1131: Vary Cache: improve encryption handling
- [x] Right now, we only check if they are defined. Should confirm their value before attempting to use them otherwise throw a warning.
- [x] Rename `set_encryption`  to `enable_encryption` (should just be a one-way function)
- [x] Move checks to `enable_encryption`
- [x] <strike>Set a random value for the key and IV if not properly defined (to avoid leaking sensitive cookie data).</strike> Throw fatal if key/iv are note properly set up.

### PR DESCRIPTION
**Pull request by dchymko**
## Description

Adds validation for
- IV/Key set if using the encrypted approach
- Cookie values only being alphanumeric, dash or underscore
- Cookie values not containing separator (refactored into single validation method)

Fixes #1131 

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).

### CODE REVIEW

**Comment by mjangda** [comment]
@dchymko I had added a few more suggested improvements in #1131 -- would be good to include those as well (namely, renaming `set_encryption` and adding a fallback for the key/iv).

**Comment by mjangda** [comment]
There's a change to the `vip-support` submodule which seems unintentional?

File: cache/class-vary-cache.php [hunk 1]
```diff
@@ -344,6 +341,21 @@ private static function set_cookie( $name, $value ) {
...
 	}
+	/**
+	 * Only allow alphanumerics, dash and underscore
+	 *
+	 * @param string $value The string you want to test on.
+	 * @return boolean
+	 */
+	private static function validate_cookie_values( $value ) {
+		if ( preg_match( '/[^_0-9a-zA-Z-]+/', $value ) > 0 ) {
+			return 'Invalid character(s). Can only use alphanumerics, dash and underscore';
```
Comment (mjangda) [inline cache/class-vary-cache.php]: Should probably return a `WP_Error` object here rather than a string?

Also, tests?

**Comment by dchymko** [comment]
I guess we may also want to validate for empty ('') values?

File: cache/class-vary-cache.php (hunk 1 above)
Comment (dchymko) [reply cache/class-vary-cache.php]: You're right, I was thinking it should be up to the calling function in how it wants to deal with the errors. But we can do:
```
if ( is_wp_error( $value) ) {
  trigger_error(...
```
in the calling functions instead.

It's tested indirectly from the tests I added in the register_groups() but I should probably add explicit tests too.

Comment (dchymko) [reply cache/class-vary-cache.php]: Done

File: cache/class-vary-cache.php [hunk 2]
```diff
@@ -64,8 +64,9 @@ public static function remove_no_cache_for_user() {
 	 */
 	public static function register_groups( array $groups ) {
 		foreach ( $groups as $group ) {
-			if ( strpos( $group, self::GROUP_SEPARATOR ) !== false || strpos( $group, self::VALUE_SEPARATOR ) !== false ) {
-				trigger_error( sprintf( 'Failed to register group (%s); cannot use the delimiter values (`%s` or `%s`) in the group name', $group, self::GROUP_SEPARATOR, self::VALUE_SEPARATOR ), E_USER_WARNING );
+			$validate_result = self::validate_cookie_values( $group );
+			if ( is_wp_error( $validate_result ) ) {
+				trigger_error( sprintf( 'Failed to register group (%s) ; ', $group, $validate_result->get_error_message() ), E_USER_WARNING );
```
Comment (mjangda) [inline cache/class-vary-cache.php]:
```suggestion
				trigger_error( sprintf( 'Failed to register group (%s): %s', $group, $validate_result->get_error_message() ), E_USER_WARNING );
```

File: cache/class-vary-cache.php [hunk 3]
```diff
@@ -116,11 +117,13 @@ public static function set_group_for_user( $group, $value ) {
...
-			return new WP_Error( 'invalid_vary_group_name', sprintf( 'Failed to register group; cannot use the delimiter values (`%s` or `%s`) in the group name', self::GROUP_SEPARATOR, self::VALUE_SEPARATOR ) );
+		$validate_group_result = self::validate_cookie_values( $group );
+		if ( is_wp_error( $validate_group_result ) ) {
+			return new WP_Error( 'invalid_vary_group_name', sprintf( 'Failed to register group (%s): %s', $group, $validate_group_result->get_error_message() ) );
 		}
-		if ( strpos( $value, self::GROUP_SEPARATOR ) !== false || strpos( $value, self::VALUE_SEPARATOR ) !== false ) {
-			return new WP_Error( 'invalid_vary_group_segment', sprintf( 'Failed to register group; cannot use the delimiter values (`%s` or `%s`) in the group segment', self::GROUP_SEPARATOR, self::VALUE_SEPARATOR ) );
+		$validate_value_result = self::validate_cookie_values( $value );
+		if ( is_wp_error( $validate_value_result ) ) {
+			return new WP_Error( 'invalid_vary_group_segment', sprintf( 'Failed to register group segment (%s); %s ', $group, $validate_value_result->get_error_message() ) );
```
Comment (mjangda) [inline cache/class-vary-cache.php]:
```suggestion
			return new WP_Error( 'invalid_vary_group_segment', sprintf( 'Failed to register group segment (%s): %s', $group, $validate_value_result->get_error_message() ) );
```

File: cache/class-vary-cache.php [hunk 4]
```diff
@@ -188,10 +191,17 @@ public static function get_groups() {
...
+	 * @return WP_Error|null
 	 */
-	public static function set_encryption( $value = true ) {
-		static::$encryption_enabled = $value;
+	public static function enable_encryption() {
+
+		// Validate that we have the secret values.
+		if ( ( ! defined( 'VIP_GO_AUTH_COOKIE_KEY' ) || ! defined( 'VIP_GO_AUTH_COOKIE_IV' ) ||
+			empty( constant( 'VIP_GO_AUTH_COOKIE_KEY' ) ) || empty( constant( 'VIP_GO_AUTH_COOKIE_IV' ) ) ) ) {
+			return new WP_Error( 'vary-cache-secrets-not-defined', sprintf( 'Constants not defined for encrypted vary cache cookies (%s and %s)', 'VIP_GO_AUTH_COOKIE_KEY', 'VIP_GO_AUTH_COOKIE_IV' ) );
```
Comment (mjangda) [inline cache/class-vary-cache.php]: Let's fatal here as discussed.

File: cache/class-vary-cache.php [hunk 5]
```diff
@@ -348,13 +344,28 @@ private static function set_cookie( $name, $value ) {
...
 	private static function unset_cookie( $name ) {
 		setcookie( $name, '', time() - 3600 );
 	}
+	/**
+	 * Only allow alphanumerics, dash and underscore
+	 *
+	 * @param string $value The string you want to test on.
+	 * @return WP_Error|boolean
+	 */
+	private static function validate_cookie_values( $value ) {
```
Comment (mjangda) [inline cache/class-vary-cache.php]: Minor: This would be better named as a singular: `validate_cookie_value` (no `s`)

File: tests/cache/test-vary-cache.php [hunk 6]
```diff
@@ -310,4 +321,69 @@ public function test__set_group_for_user_invalid( $group, $value, $expected_erro
...
+
+		$actual_result = Vary_Cache::enable_encryption( );
+
+		$this->assertWPError( $actual_result, 'Not WP_Error object' );
+		$actual_error_code = $actual_result->get_error_code();
+		$this->assertEquals( 'vary-cache-secrets-not-defined', $actual_error_code, 'Incorrect error code' );
+
+	}
+
+	public function test__enable_encryption_true_valid() {
```
Comment (mjangda) [inline tests/cache/test-vary-cache.php]: Will need to run this test in a separate process otherwise the `define` will persist across other tests.

See https://github.com/Automattic/vip-go-mu-plugins/blob/2280341588b2fe2fff26fec8b4cce984d2aa27df/tests/lib/utils/context.php#L40-L43

File: tests/cache/test-vary-cache.php [hunk 7]
```diff
@@ -310,4 +321,69 @@ public function test__set_group_for_user_invalid( $group, $value, $expected_erro
 		$this->assertEquals( $expected_error_code, $actual_error_code, 'Incorrect error code' );
 	}
 
+	public function test__enable_encryption_true_invalid() {
```
Comment (mjangda) [inline tests/cache/test-vary-cache.php]: Also need a test for empty constants

File: tests/cache/test-vary-cache.php (hunk 6 above)
Comment (dchymko) [reply tests/cache/test-vary-cache.php]: Cool, was wondering on that

**Comment by dchymko** [comment]
This should be good to go. Note, I had to skip the enable_encryption error-checking tests due to an incompatibility in the annotation of Error exceptions between the local version of phpunit (6.5.13) and the Travis version (5.7). We may want to bump phpunit on Travis if possible

**Comment by dchymko** [comment]
r132024-deploy
